### PR TITLE
Prevent silent failures when two "databaseChangeLog" tags are present

### PR DESCRIPTION
**Pull request by danielthegray**
## Environment
**Liquibase Version**: all

**Liquibase Integration & Version**: all

**Liquibase Extension(s) & Version**: all

**Database Vendor & Version**: all

**Operating System Type & Version**: all

## Pull Request Type

* [ ] Bug fix (non-breaking change which fixes an issue.)
* [x] Enhancement/New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
If someone makes a mistake and copy/pastes a databaseChangeLog block in,
and makes the file have two tags, the SnakeYAML parser silently covers
up the first instance with the second one, which will cause the first
block of changes to be silently omitted.

This can be very problematic and cause hard-to-debug errors, which is
why I have added a small FilterInputStream interceptor to check the
stream as it is fed into the YAML parser, for a duplicate tag. Doing it
with the FilterInputStream like this prevents the efficiency losses of
opening the file twice.

## Steps To Reproduce
You can look at the added test; essentially, just create a changelog file which has two "databaseChangeLog" tags.

## Actual Behavior
It will silently to the wrong thing (i.e. skip over the first group of changesets).

## Expected/Desired Behavior
An error should be thrown to highlight the fact that there are two databaseChangeLog tags/keys.

## Fast Track PR Acceptance Checklist:

* [x] Build is successful and all new and existing tests pass
* [x] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
* [ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
* [ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
* [ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

### CODE REVIEW

**Comment by asomov** [comment]
@danielthegray SnakeYAML has a property to [prohibit the same keys](https://bitbucket.org/snakeyaml/snakeyaml/issues/395/duplicate-key-diagnostics)

**Comment by daniel-gray-jemmic** [comment]
> @danielthegray SnakeYAML has a property to [prohibit the same keys](https://bitbucket.org/snakeyaml/snakeyaml/issues/395/duplicate-key-diagnostics)

Thank you @asomov for the great tip, it greatly simplifies the code! I've pushed a new revision of the branch with conflicts resolved, but the "FOSSA License Compliance and Security Check" is failing. Does something need to be added to the branch code?

**Comment by filipelautert** [comment]
@daniel-gray-jemmic currently fossa is failing for contributed PR's, don't worry about it.

**Review by filipelautert — APPROVED** [review]

File: liquibase-standard/src/test/groovy/liquibase/parser/core/yaml/YamlChangeLogParser_RealFile_Test.groovy [hunk 1]
```diff
@@ -299,6 +299,15 @@ public class YamlChangeLogParser_RealFile_Test extends Specification {
         assert e.message.startsWith("Syntax error in file liquibase/parser/core/yaml/malformedChangeLog.yaml")
     }
 
+    def "ChangeLogParseException thrown if changelog has two databaseChangeLog tags"() throws Exception {
+        when:
+        new YamlChangeLogParser().parse("liquibase/parser/core/yaml/malformedDoubleChangeLog.yaml", new ChangeLogParameters(), new JUnitResourceAccessor())
+
+        then:
+        def e = thrown(ChangeLogParseException)
+        assert e.message.startsWith("Syntax error in file liquibase/parser/core/yaml/malformedDoubleChangeLog.yaml")
```
Comment (StevenMassaro) [inline liquibase-standard/src/test/groovy/liquibase/parser/core/yaml/YamlChangeLogParser_RealFile_Test.groovy:308]: Does this error message contain more useful information in it?

**Review by StevenMassaro — APPROVED** [review]

File: liquibase-standard/src/test/groovy/liquibase/parser/core/yaml/YamlChangeLogParser_RealFile_Test.groovy (hunk 1 above)
Comment (daniel-gray-jemmic) [reply liquibase-standard/src/test/groovy/liquibase/parser/core/yaml/YamlChangeLogParser_RealFile_Test.groovy:308]: Indeed, it does, I've added an additional assertion on the message. Thanks for the great idea!

**Review by StevenMassaro — APPROVED** [review]

**Review by rberezen — APPROVED** [review]

**Comment by MalloD12** [comment]
Hi @daniel-gray-jemmic,

I just wanted to let you know today I commented out the line you added on the `YamlParser` class, plus ignored the test you also added in this PR because this change was breaking the case of having a duplicate `sql:` property. I created the issue #5963  in order to fine a way to keep using your code change, but without breaking some other case like the one mentioned in that issue. 

If you have any ideas or would like to work on it, please feel free to assign it to yourself or ask any questions/concerns you might have.

Thank you and apologies for this,
Daniel.

**Comment by filipelautert** [comment]
> Hi, I understand. The first commit I had on this issue was ONLY detecting the duplicate databaseChangeLog, and I changed it to the duplicate flag because of a suggestion in a comment on this thread. Should we try to revert to the previous version instead?

@danielthegray I think his would be a better approach, thanks!

**Pull request merged**